tajo-catalog/tajo-catalog-server/src/main/java/org/apache/tajo/catalog/store/AbstractDBStore.java

```diff
@@ -2228,6 +2238,260 @@ private void setPartitionKeys(int pid, PartitionDescProto.Builder partitionDesc)
     return partitions;
   }
 
+  /**
+   * Check if list of partitions exist on catalog.
+   *
+   *
+   * @param databaseId
+   * @param tableId
+   * @return
+   */
+  public boolean existPartitionsOnCatalog(int tableId) {
+    Connection conn = null;
+    ResultSet res = null;
+    PreparedStatement pstmt = null;
+    boolean result = false;
+
+    try {
+      String sql = "SELECT * FROM " + TB_PARTTIONS +" WHERE " + COL_TABLES_PK + " = ?  ";
+      if (LOG.isDebugEnabled()) {
+        LOG.debug(sql);
+      }
+
+      conn = getConnection();
+      pstmt = conn.prepareStatement(sql);
+      pstmt.setInt(1, tableId);
+      res = pstmt.executeQuery();
+
+      if (res.next()) {
+        result = true;
+      }
+    } catch (SQLException se) {
+      throw new TajoInternalError(se);
+    } finally {
+      CatalogUtil.closeQuietly(pstmt, res);
+    }
+    return result;
+  }
+
+  @Override
+  public List<PartitionDescProto> getPartitionsByFilter(PartitionsByFilterProto request) {
+    throw new TajoRuntimeException(new UnsupportedException());
+  }
+
+  @Override
+  public List<PartitionDescProto> getPartitionsByAlgebra(PartitionsByAlgebraProto request) throws
+      UndefinedDatabaseException, UndefinedTableException, UndefinedPartitionMethodException,
+      PartitionNotFoundException, UnsupportedException {
+    Connection conn = null;
+    PreparedStatement pstmt = null;
+    ResultSet res = null;
+    int currentIndex = 1;
+    String selectStatement = null;
+    Pair<String, List<PartitionFilterSet>> pair = null;
+
+    List<PartitionDescProto> partitions = TUtil.newList();
+    List<PartitionFilterSet> filterSets = null;
+
+    int databaseId = getDatabaseId(request.getDatabaseName());
+    int tableId = getTableId(databaseId, request.getDatabaseName(), request.getTableName());
+    if (!existPartitionMethod(request.getDatabaseName(), request.getTableName())) {
+      throw new UndefinedPartitionMethodException(request.getTableName());
+    }
+
+    if (!existPartitionsOnCatalog(tableId)) {
+      throw new PartitionNotFoundException(request.getTableName());
```

Comment:
```CatalogStore::getPartitionsOfTable``` can returns empty list instead of ```PartitionNotFoundException```. But if ```CatalogStore::getPartitionsByAlgebra``` doesn't return ```PartitionNotFoundException```, ```PartitionedTableRewriter``` might not check in case of hdfs listing need.